File: security/src/main/java/top/bettercode/summer/security/ScopeUserDetailsService.java
package top.bettercode.summer.security;

import java.util.Objects;
import javax.servlet.http.HttpServletRequest;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.util.Assert;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

/**
 * @author Peter Wu
 */
public interface ScopeUserDetailsService extends UserDetailsService {

  UserDetails loadUserByScopeAndUsername(String scope, String username)
      throws UsernameNotFoundException;

  @Override
  default UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
    ServletRequestAttributes requestAttributes = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
    HttpServletRequest request = Objects.requireNonNull(requestAttributes).getRequest();
    String scope = request.getParameter(SecurityParameterNames.SCOPE);
    Assert.hasText(scope, "scope 不能为空");
    return loadUserByScopeAndUsername(scope, username);
  }
}

File: security/src/main/kotlin/top/bettercode/summer/security/userdetails/ScopeUserDetailsService.kt
package top.bettercode.summer.security.userdetails

import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.util.Assert
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import top.bettercode.summer.security.support.SecurityParameterNames

/**
 * @author Peter Wu
 */
interface ScopeUserDetailsService : UserDetailsService {
    fun loadUserByScopeAndUsername(scope: String, username: String): UserDetails

    override fun loadUserByUsername(username: String): UserDetails {
        val requestAttributes = RequestContextHolder.getRequestAttributes() as ServletRequestAttributes
        val request = requestAttributes.request
        val scope = request.getParameter(SecurityParameterNames.SCOPE)
        Assert.hasText(scope, "scope 不能为空")
        return loadUserByScopeAndUsername(scope, username)
    }
}
